libs/common/Networking/TcpNetworkHandlerBase.cs

```diff
@@ -135,31 +136,52 @@ void Dispose(SocketAsyncEventArgs e)
         }
 
         void RecvEventArg_Completed(object sender, SocketAsyncEventArgs e)
+        {
+            if (e.BytesTransferred == 0 || e.SocketError != SocketError.Success || serverHook.Disposed)
+            {
+                // No more things to receive
+                Dispose(e);
+                return;
+            }
+            // Complete receive event and release thread while we process data async
+            _ = HandleReceiveAsync(sender, e);
+        }
+
+        private async ValueTask HandleReceiveAsync(object sender, SocketAsyncEventArgs e)
         {
             try
             {
-                do
+                var receiveTask = OnNetworkReceiveAsync(e.BytesTransferred);
+                if (!receiveTask.IsCompleted)
                 {
-                    if (e.BytesTransferred == 0 || e.SocketError != SocketError.Success || serverHook.Disposed)
-                    {
-                        // No more things to receive
-                        Dispose(e);
-                        break;
-                    }
-                    OnNetworkReceive(e.BytesTransferred);
-                    e.SetBuffer(networkReceiveBuffer, networkBytesRead, networkReceiveBuffer.Length - networkBytesRead);
-                } while (!e.AcceptSocket.ReceiveAsync(e));
+                    await receiveTask;
+                }
+                NetworkBufferProcessed(sender, e);
             }
             catch (Exception ex)
             {
-                if (ex is ObjectDisposedException ex2 && ex2.ObjectName == "System.Net.Sockets.Socket")
-                    logger?.LogTrace("Accept socket was disposed at RecvEventArg_Completed");
-                else
-                    logger?.LogError(ex, "An error occurred at RecvEventArg_Completed");
-                Dispose(e);
+                HandleReceiveFailure(ex, e);
             }
         }
 
+        void NetworkBufferProcessed(object sender, SocketAsyncEventArgs e)
+        {
+            e.SetBuffer(networkReceiveBuffer, networkBytesRead, networkReceiveBuffer.Length - networkBytesRead);
+            if (!e.AcceptSocket.ReceiveAsync(e))
```

Comment:
should this be a while loop?

Reply:
this was being handled in continuation but maybe that might create extra stack frame if we get lots of sync completions. Replacing it with do while (as before) since it might be more optimal. Thanks!